In [58]:
import pandas as pd 
import numpy as np 
import os 

print(os.getcwd())

/Users/eugenechua/Downloads/course_skill_proficiency_evaluator


In [37]:
# Reading in the course skill extractions.
course_skill_df = pd.read_excel('./data/final_course_skills_merged.xlsx')
print(course_skill_df.shape)
print(course_skill_df['coursereferencenumber'].nunique())
course_skill_df.head(5)

(916, 12)
794


,coursereferencenumber,coursetitle,coursedescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,uniqueskilltitle,proficiencylevelofcourse
0,TGS-2019503800,Combi Oven Cooking for Chinese Cuisine,"In this hands-on workshop, participants will l...",Chinese Dry-Heat Dish Preparation and Cooking,tsc,Intermediate,0.8737,0.677110,1,full text,Chinese Dry-Heat Dish Preparation and Cooking,Basic
1,TGS-2019503804,Social Media for Food Marketing,Particpants will examine the use of social med...,Social Media Marketing,tsc,Basic,0.7964,0.731026,1,full text,Social Media Marketing,Basic
2,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8033,0.890494,1,data governance,Data Governance,Intermediate
3,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8406,0.694061,1,full text,Data Strategy,Intermediate
4,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8406,0.694061,1,full text,Data Strategy,Intermediate


In [4]:
course_skill_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 916 entries, 0 to 915
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   coursereferencenumber     916 non-null    object 
 1   coursetitle               916 non-null    object 
 2   coursedescription         916 non-null    object 
 3   skillTitle                916 non-null    object 
 4   skillSource               916 non-null    object 
 5   proficiencyLevel          916 non-null    object 
 6   proficiencyScore          871 non-null    float64
 7   confidenceScore           916 non-null    float64
 8   skillOccurrence           916 non-null    int64  
 9   phrase                    915 non-null    object 
 10  uniqueskilltitle          916 non-null    object 
 11  proficiencylevelofcourse  916 non-null    object 
dtypes: float64(2), int64(1), object(9)
memory usage: 86.0+ KB


In [5]:
course_skill_df.isnull().sum()

coursereferencenumber        0
coursetitle                  0
coursedescription            0
skillTitle                   0
skillSource                  0
proficiencyLevel             0
proficiencyScore            45
confidenceScore              0
skillOccurrence              0
phrase                       1
uniqueskilltitle             0
proficiencylevelofcourse     0
dtype: int64

In [6]:
# Reading in course skill with updated algo
course_skill_updated_algo_df = pd.read_excel('./data/final_course_skills_merged_modified_algo.xlsx')
print(course_skill_updated_algo_df.shape)
print(course_skill_updated_algo_df['coursereferencenumber'].nunique())
course_skill_updated_algo_df.head(5)

(916, 12)
794


,coursereferencenumber,coursetitle,coursedescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,uniqueskilltitle,proficiencylevelofcourse
0,TGS-2019503800,Combi Oven Cooking for Chinese Cuisine,"In this hands-on workshop, participants will l...",Chinese Dry-Heat Dish Preparation and Cooking,tsc,Intermediate,0.8566,0.677110,1,full text,Chinese Dry-Heat Dish Preparation and Cooking,Basic
1,TGS-2019503804,Social Media for Food Marketing,Particpants will examine the use of social med...,Social Media Marketing,tsc,Advanced,0.8116,0.731026,1,full text,Social Media Marketing,Basic
2,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8171,0.890494,1,data governance,Data Governance,Intermediate
3,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8295,0.694061,1,full text,Data Strategy,Intermediate
4,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8295,0.694061,1,full text,Data Strategy,Intermediate


In [7]:
course_skill_updated_algo_df.isnull().sum()

coursereferencenumber        0
coursetitle                  0
coursedescription            0
skillTitle                   0
skillSource                  0
proficiencyLevel             0
proficiencyScore            45
confidenceScore              0
skillOccurrence              0
phrase                       1
uniqueskilltitle             0
proficiencylevelofcourse     0
dtype: int64

In [8]:
course_skill_updated_algo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 916 entries, 0 to 915
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   coursereferencenumber     916 non-null    object 
 1   coursetitle               916 non-null    object 
 2   coursedescription         916 non-null    object 
 3   skillTitle                916 non-null    object 
 4   skillSource               916 non-null    object 
 5   proficiencyLevel          916 non-null    object 
 6   proficiencyScore          871 non-null    float64
 7   confidenceScore           916 non-null    float64
 8   skillOccurrence           916 non-null    int64  
 9   phrase                    915 non-null    object 
 10  uniqueskilltitle          916 non-null    object 
 11  proficiencylevelofcourse  916 non-null    object 
dtypes: float64(2), int64(1), object(9)
memory usage: 86.0+ KB


In [59]:
# Reading in the course skills with chatgpt proficiency labelling to see what it looks like
course_skill_gpt_labelled = pd.read_csv('./data/courses_skills_with_proficiency_labels.csv')
print(course_skill_gpt_labelled.shape)
print(course_skill_gpt_labelled['coursereferencenumber'].nunique())
course_skill_gpt_labelled.head(10)

(916, 16)
794


,coursereferencenumber,coursetitle,coursedescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,uniqueskilltitle,proficiencylevelofcourse,gpt_proficiencyLevel,gpt_rationale,gpt_raw_response,gpt_error
0,TGS-2019503800,Combi Oven Cooking for Chinese Cuisine,"In this hands-on workshop, participants will l...",Chinese Dry-Heat Dish Preparation and Cooking,tsc,Intermediate,0.8737,0.677110,1,full text,Chinese Dry-Heat Dish Preparation and Cooking,Basic,Basic,The course description states that participant...,"{\n ""skillTitle"": ""Chinese Dry-Heat Dish Prep...",NaN
1,TGS-2019503804,Social Media for Food Marketing,Particpants will examine the use of social med...,Social Media Marketing,tsc,Basic,0.7964,0.731026,1,full text,Social Media Marketing,Basic,Intermediate,The course description uses the word 'examine'...,"{\n ""skillTitle"": ""Social Media Marketing"",\n...",NaN
2,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8033,0.890494,1,data governance,Data Governance,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Governance"",\n ""prof...",NaN
3,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8406,0.694061,1,full text,Data Strategy,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Strategy"",\n ""profic...",NaN
4,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8406,0.694061,1,full text,Data Strategy,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Strategy"",\n ""profic...",NaN
5,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8033,0.890494,1,data governance,Data Governance,Intermediate,Basic,The course description focuses on learning abo...,"{\n ""skillTitle"": ""Data Governance"",\n ""prof...",NaN
6,TGS-2020501272,Train The Trainer - Develop and Conduct Assess...,•\tFormulate assessment scheme\r\n•\tDevel...,Assessment Design and Implementation,tsc,Intermediate,0.7604,0.797698,2,assessment scheme Develop theory assessment sp...,Assessment Design and Implementation,Intermediate,Advanced,The course requires learners to 'formulate ass...,"{\n ""skillTitle"": ""Assessment Design and Impl...",NaN
7,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Applications Development,tsc,Intermediate,0.5980,0.609969,1,full text,Applications Development,Intermediate,Basic,The course introduces foundational concepts an...,"{\n ""skillTitle"": ""Applications Development"",...",NaN
8,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Software Testing,tsc,Advanced,0.6417,0.651658,1,testing,Software Testing,Intermediate,Basic,The course description states students will be...,"{\n ""skillTitle"": ""Software Testing"",\n ""pro...",NaN
9,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Agile Software Development,tsc,Basic,0.6478,0.608859,3,software development practices,Agile Software Development,Intermediate,Basic,The course description states that students wi...,"{\n ""skillTitle"": ""Agile Software Development...",NaN


In [46]:
print(course_skill_gpt_labelled['gpt_proficiencyLevel'].value_counts())
print(course_skill_gpt_labelled['gpt_proficiencyLevel'].value_counts(normalize=True))

gpt_proficiencyLevel
Intermediate    418
Basic           364
Advanced        134
Name: count, dtype: int64
gpt_proficiencyLevel
Intermediate    0.456332
Basic           0.397380
Advanced        0.146288
Name: proportion, dtype: float64


In [47]:
print(course_skill_gpt_labelled['proficiencyLevel'].value_counts())
print(course_skill_gpt_labelled['proficiencyLevel'].value_counts(normalize=True))

proficiencyLevel
Intermediate    424
Advanced        302
Basic           190
Name: count, dtype: int64
proficiencyLevel
Intermediate    0.462882
Advanced        0.329694
Basic           0.207424
Name: proportion, dtype: float64


### Performing model evaluation now on both baseline and slightly modified models

In [48]:
# Doing validation for the baseline algo (non modified!)
from sklearn.metrics import f1_score, precision_score, classification_report

y_pred = np.array(course_skill_gpt_labelled['proficiencyLevel'])
y_true = np.array(course_skill_gpt_labelled['gpt_proficiencyLevel'])

# Performing encoding of labels!
label_map = {'Basic': 0, 'Intermediate': 1, 'Advanced': 2}
y_pred = np.array([label_map[l] for l in y_pred])
y_true = np.array([label_map[l] for l in y_true])

print(y_pred[:20])
print(y_true[:20])


[1 0 2 2 2 2 1 1 2 0 0 2 2 2 2 2 2 0 0 0]
[0 1 0 0 0 0 2 0 0 0 0 2 1 2 2 2 2 1 1 1]


In [49]:
f1 = f1_score(y_true, y_pred, average='weighted', zero_division=0)
precision = precision_score(y_true, y_pred, average='weighted', zero_division=0)

print(f'weighted f1 score is {f1}')
print(f'weighted precision score is {precision}')

weighted f1 score is 0.35967325273771955
weighted precision score is 0.40333682398978576


In [50]:
y_pred = np.array(course_skill_gpt_labelled['proficiencyLevel'])
y_true = np.array(course_skill_gpt_labelled['gpt_proficiencyLevel'])

print(classification_report(y_true, y_pred, target_names=['Basic', 'Intermediate', 'Advanced'], labels=['Basic', 'Intermediate', 'Advanced']))


              precision    recall  f1-score   support

       Basic       0.43      0.22      0.29       364
Intermediate       0.47      0.47      0.47       418
    Advanced       0.14      0.32      0.20       134

    accuracy                           0.35       916
   macro avg       0.35      0.34      0.32       916
weighted avg       0.40      0.35      0.36       916



In [60]:
# Reading in the course skills with chatgpt proficiency labelling (modified algo) to see what it looks like
course_skill_gpt_labelled = pd.read_csv('./data/courses_skills_with_proficiency_labels_modified_algo.csv')
print(course_skill_gpt_labelled.shape)
print(course_skill_gpt_labelled['coursereferencenumber'].nunique())
course_skill_gpt_labelled.head(10)

(916, 16)
794


,coursereferencenumber,coursetitle,coursedescription,skillTitle,skillSource,proficiencyLevel,proficiencyScore,confidenceScore,skillOccurrence,phrase,uniqueskilltitle,proficiencylevelofcourse,gpt_proficiencyLevel,gpt_rationale,gpt_raw_response,gpt_error
0,TGS-2019503800,Combi Oven Cooking for Chinese Cuisine,"In this hands-on workshop, participants will l...",Chinese Dry-Heat Dish Preparation and Cooking,tsc,Intermediate,0.8566,0.677110,1,full text,Chinese Dry-Heat Dish Preparation and Cooking,Basic,Basic,The course description states participants wil...,"{\n ""skillTitle"": ""Chinese Dry-Heat Dish Prep...",NaN
1,TGS-2019503804,Social Media for Food Marketing,Particpants will examine the use of social med...,Social Media Marketing,tsc,Advanced,0.8116,0.731026,1,full text,Social Media Marketing,Basic,Intermediate,The course description states that participant...,"{\n ""skillTitle"": ""Social Media Marketing"",\n...",NaN
2,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8171,0.890494,1,data governance,Data Governance,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Governance"",\n ""prof...",NaN
3,TGS-2020000390,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8295,0.694061,1,full text,Data Strategy,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Strategy"",\n ""profic...",NaN
4,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Strategy,tsc,Advanced,0.8295,0.694061,1,full text,Data Strategy,Intermediate,Basic,The course description focuses on identifying ...,"{\n ""skillTitle"": ""Data Strategy"",\n ""profic...",NaN
5,TGS-2020001034,"Advanced Certificate in Governance, Risk Manag...",All organisations need to plan how they use da...,Data Governance,tsc,Advanced,0.8171,0.890494,1,data governance,Data Governance,Intermediate,Basic,The course description focuses on learning abo...,"{\n ""skillTitle"": ""Data Governance"",\n ""prof...",NaN
6,TGS-2020501272,Train The Trainer - Develop and Conduct Assess...,•\tFormulate assessment scheme\r\n•\tDevel...,Assessment Design and Implementation,tsc,Advanced,0.7710,0.797698,2,assessment scheme Develop theory assessment sp...,Assessment Design and Implementation,Intermediate,Advanced,The course requires learners to 'formulate ass...,"{\n ""skillTitle"": ""Assessment Design and Impl...",NaN
7,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Agile Software Development,tsc,Intermediate,0.6915,0.608859,3,software development practices,Agile Software Development,Intermediate,Basic,The course description states that students wi...,"{\n ""skillTitle"": ""Agile Software Development...",NaN
8,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Software Testing,tsc,Intermediate,0.7392,0.651658,1,testing,Software Testing,Intermediate,Basic,The course description states students will be...,"{\n ""skillTitle"": ""Software Testing"",\n ""pro...",NaN
9,TGS-2020502315,Software Development Fundamentals,This course aims to introduce non-computing st...,Software Design,tsc,Basic,0.7077,0.711091,3,software development,Software Design,Intermediate,Basic,"The course introduces principles and concepts,...","{\n ""skillTitle"": ""Software Design"",\n ""prof...",NaN


In [57]:
print(course_skill_gpt_labelled['proficiencyLevel'].value_counts())
print(course_skill_gpt_labelled['proficiencyLevel'].value_counts(normalize=True))

proficiencyLevel
Intermediate    452
Advanced        290
Basic           174
Name: count, dtype: int64
proficiencyLevel
Intermediate    0.493450
Advanced        0.316594
Basic           0.189956
Name: proportion, dtype: float64


In [52]:
print(course_skill_gpt_labelled['gpt_proficiencyLevel'].value_counts())
print(course_skill_gpt_labelled['gpt_proficiencyLevel'].value_counts(normalize=True))

gpt_proficiencyLevel
Intermediate    418
Basic           365
Advanced        133
Name: count, dtype: int64
gpt_proficiencyLevel
Intermediate    0.456332
Basic           0.398472
Advanced        0.145197
Name: proportion, dtype: float64


In [53]:
# Doing validation for the MODIFIED baseline algo!)
from sklearn.metrics import f1_score, precision_score, classification_report

y_pred = np.array(course_skill_gpt_labelled['proficiencyLevel'])
y_true = np.array(course_skill_gpt_labelled['gpt_proficiencyLevel'])

# Performing encoding of labels!
label_map = {'Basic': 0, 'Intermediate': 1, 'Advanced': 2}
y_pred = np.array([label_map[l] for l in y_pred])
y_true = np.array([label_map[l] for l in y_true])

print(y_pred[:20])
print(y_true[:20])

[1 2 2 2 2 2 2 1 1 0 1 2 2 2 1 2 2 0 0 0]
[0 1 0 0 0 0 2 0 0 0 0 2 1 2 2 2 2 1 1 1]


In [54]:
f1 = f1_score(y_true, y_pred, average='weighted', zero_division=0)
precision = precision_score(y_true, y_pred, average='weighted', zero_division=0)

print(f'weighted f1 score is {f1}')
print(f'weighted precision score is {precision}')

weighted f1 score is 0.36855332060525553
weighted precision score is 0.4221563639481028


In [55]:
y_pred = np.array(course_skill_gpt_labelled['proficiencyLevel'])
y_true = np.array(course_skill_gpt_labelled['gpt_proficiencyLevel'])

print(classification_report(y_true, y_pred, target_names=['Basic', 'Intermediate', 'Advanced'], labels=['Basic', 'Intermediate', 'Advanced']))


              precision    recall  f1-score   support

       Basic       0.49      0.23      0.32       365
Intermediate       0.46      0.49      0.47       418
    Advanced       0.13      0.29      0.18       133

    accuracy                           0.36       916
   macro avg       0.36      0.34      0.32       916
weighted avg       0.42      0.36      0.37       916



Alright the performances of the 2 baseline model variants is quite abit underwhelming ~ 40% - 42% precision is not quite acceptable..In the meantime, I will try testing the SEA with PL extractor on job postings descriptions, before moving on to think about tweaking the model architecture.

The fundamental problem is that course description are quite often badly written. There are numerous instances where the course title states its advanced, but the description are in bullet points and very basic, hence making it impossibly hard for the PL extractor to even extract anything relevant..